aioshelly/block_device/coap.py

```diff
@@ -36,9 +39,34 @@ def __init__(self, sender_addr: tuple[str, int], payload: bytes) -> None:
         if self.code not in (30, 69):
             raise InvalidMessage(f"Wrong type, {self.code}")
 
+        raw_data = payload[4:]
+        option_number = 0
+        data = b""
+
+        # parse options
+        while raw_data:
+            if raw_data[0] == 0xFF:  # end of options marker
+                data = raw_data[1:]
+                break
+
+            delta = (raw_data[0] & 0xF0) >> 4
+            length = raw_data[0] & 0x0F
+            (delta, raw_data) = self._read_extended_field_value(delta, raw_data[1:])
+            (length, raw_data) = self._read_extended_field_value(length, raw_data)
+            option_number += delta
+
+            if len(raw_data) < length:
+                raise InvalidMessage("Option announced but absent")
+
+            self.options[option_number] = raw_data[:length]
+            raw_data = raw_data[length:]
+
+        if not data:
+            raise InvalidMessage("Received message without data")
+
         try:
-            self.payload = json.loads(payload.rsplit(b"\xff", 1)[1].decode())
-        except (json.decoder.JSONDecodeError, UnicodeDecodeError, IndexError) as err:
+            self.payload = json.loads(data.decode())
```

Comment:
```suggestion
            self.payload = json_loads(data.decode())
```

Also needs
`from ..json import json_dumps, json_loads`